File: ios/AdPlay/Stores/PlayerSettings.swift
import Foundation
import Combine

@MainActor
final class PlayerSettings: ObservableObject {
    @Published var remindersEnabled: Bool {
        didSet { UserDefaults.standard.set(remindersEnabled, forKey: Keys.reminders) }
    }
    @Published var hapticsEnabled: Bool {
        didSet { UserDefaults.standard.set(hapticsEnabled, forKey: Keys.haptics) }
    }
    @Published var soundEnabled: Bool {
        didSet { UserDefaults.standard.set(soundEnabled, forKey: Keys.sound) }
    }
    @Published var hasCompletedOnboarding: Bool {
        didSet { UserDefaults.standard.set(hasCompletedOnboarding, forKey: Keys.onboarding) }
    }

    @Published private(set) var seenDailyGoalDay: String
    @Published private(set) var seenDailyGoalIds: [String]

    init() {
        let d = UserDefaults.standard
        remindersEnabled = d.object(forKey: Keys.reminders) as? Bool ?? true
        hapticsEnabled = d.object(forKey: Keys.haptics) as? Bool ?? true
        soundEnabled = d.object(forKey: Keys.sound) as? Bool ?? true
        hasCompletedOnboarding = d.bool(forKey: Keys.onboarding)
        seenDailyGoalDay = d.string(forKey: Keys.seenGoalDay) ?? ""
        seenDailyGoalIds = d.stringArray(forKey: Keys.seenGoalIds) ?? []
    }

    func unseenCompletedGoalCount(in goals: [DailyGoal]) -> Int {
        let seen = seenDailyGoalDay == Self.utcDayKey() ? Set(seenDailyGoalIds) : []
        return goals.filter { $0.completed && !seen.contains($0.id) }.count
    }

    func acknowledgeDailyGoals(_ goals: [DailyGoal]) {
        seenDailyGoalDay = Self.utcDayKey()
        seenDailyGoalIds = goals.filter(\.completed).map(\.id)
        let d = UserDefaults.standard
        d.set(seenDailyGoalDay, forKey: Keys.seenGoalDay)
        d.set(seenDailyGoalIds, forKey: Keys.seenGoalIds)
    }

    static func utcDayKey() -> String {
        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = TimeZone(identifier: "UTC")!
        let c = cal.dateComponents([.year, .month, .day], from: Date())
        return String(format: "%04d-%02d-%02d", c.year ?? 0, c.month ?? 0, c.day ?? 0)
    }

    private enum Keys {
        static let reminders = "adplay.remindersEnabled"
        static let haptics = "adplay.hapticsEnabled"
        static let sound = "adplay.soundEnabled"
        static let onboarding = "adplay.hasCompletedOnboarding"
        static let seenGoalDay = "adplay.seenDailyGoalDay"
        static let seenGoalIds = "adplay.seenDailyGoalIds"
    }
}

File: android/app/src/main/java/com/adplay/app/data/PlayerSettings.kt
package com.adplay.app.data

import android.content.Context

class PlayerSettings(context: Context) {
    private val prefs = context.applicationContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    var remindersEnabled: Boolean
        get() = prefs.getBoolean(REMINDERS, true)
        set(value) { prefs.edit().putBoolean(REMINDERS, value).apply() }

    var hapticsEnabled: Boolean
        get() = prefs.getBoolean(HAPTICS, true)
        set(value) { prefs.edit().putBoolean(HAPTICS, value).apply() }

    var soundEnabled: Boolean
        get() = prefs.getBoolean(SOUND, true)
        set(value) { prefs.edit().putBoolean(SOUND, value).apply() }

    var hasCompletedOnboarding: Boolean
        get() = prefs.getBoolean(ONBOARDING, false)
        set(value) { prefs.edit().putBoolean(ONBOARDING, value).apply() }

    fun unseenCompletedGoalCount(goals: List<DailyGoal>): Int {
        val seen = if (seenDay() == utcDayKey()) seenIds() else emptySet()
        return goals.count { it.completed && it.id !in seen }
    }

    fun acknowledgeDailyGoals(goals: List<DailyGoal>) {
        prefs.edit()
            .putString(SEEN_GOAL_DAY, utcDayKey())
            .putStringSet(SEEN_GOAL_IDS, goals.filter { it.completed }.map { it.id }.toSet())
            .apply()
    }

    private fun seenDay(): String = prefs.getString(SEEN_GOAL_DAY, "") ?: ""

    private fun seenIds(): Set<String> = prefs.getStringSet(SEEN_GOAL_IDS, emptySet()) ?: emptySet()

    companion object {
        private const val PREFS = "adplay_player"
        const val REMINDERS = "remindersEnabled"
        private const val HAPTICS = "hapticsEnabled"
        private const val SOUND = "soundEnabled"
        private const val ONBOARDING = "hasCompletedOnboarding"
        private const val SEEN_GOAL_DAY = "seenDailyGoalDay"
        private const val SEEN_GOAL_IDS = "seenDailyGoalIds"

        fun utcDayKey(): String {
            val cal = java.util.Calendar.getInstance(java.util.TimeZone.getTimeZone("UTC"))
            return String.format(
                "%04d-%02d-%02d",
                cal.get(java.util.Calendar.YEAR),
                cal.get(java.util.Calendar.MONTH) + 1,
                cal.get(java.util.Calendar.DAY_OF_MONTH),
            )
        }
    }
}
